File: app/src/main/java/com/android/orion/utility/Market.java
package com.android.orion.utility;

import java.util.Calendar;

public class Market {
    private static final String FIRST_HALF_START_TIME = "09:30:00";
    private static final String FIRST_HALF_END_TIME = "11:30:00";
    private static final String SECOND_HALF_START_TIME = "13:00:00";
    public static final String SECOND_HALF_END_TIME = "15:00:00";

    private static final int START_IN_MINUTES = 9 * 60 + 30; //FIRST_HALF_START_TIME
    private static final int LUNCH_TIME_IN_MINUTES = 1 * 60 + 30; //from FIRST_HALF_END_TIME to SECOND_HALF_START_TIME

    private static int getMinutesOfToday() {
        Calendar calendar = Calendar.getInstance();
        return (calendar.get(Calendar.HOUR_OF_DAY) * 60
                + calendar.get(Calendar.MINUTE));
    }

    public static int getScheduleMinutes() {
        if (!isWeekday()) {
            return 0;
        } else if (isFirstHalf()) {
            return getMinutesOfToday() - START_IN_MINUTES;
        } else if (isSecondHalf()) {
            return getMinutesOfToday() - (START_IN_MINUTES
                    + LUNCH_TIME_IN_MINUTES);
        } else {
            return 0;
        }
    }

    public static boolean isWeekday() {
        Calendar calendar = Calendar.getInstance();
        int dayOfWeek = calendar.get(Calendar.DAY_OF_WEEK);
        return dayOfWeek > Calendar.SUNDAY && dayOfWeek < Calendar.SATURDAY;
    }

    public static boolean isOutofDate(String modified) {
        boolean result = false;
        if (modified == null || modified.isEmpty()) {
            result = true;
        }
        if (!modified.contains(Utility.getCalendarDateString(Calendar.getInstance()))) {
            result = true;
        }
        return result;
    }

    public static boolean isTradingHours() {
        return isFirstHalf() || isSecondHalf();
    }

    public static boolean beforeOpen() {
        Calendar calendar = Calendar.getInstance();
        if (!isWeekday()) {
            return false;
        } else {
            return calendar.before(getFirstHalfStartCalendar());
        }
    }

    public static boolean isFirstHalf() {
        Calendar calendar = Calendar.getInstance();
        if (!isWeekday()) {
            return false;
        } else {
            return calendar.after(getFirstHalfStartCalendar())
                    && calendar.before(getFirstHalfEndCalendar());
        }
    }

    public static boolean isLunchTime() {
        Calendar calendar = Calendar.getInstance();
        if (!isWeekday()) {
            return false;
        } else {
            return calendar.after(getFirstHalfEndCalendar())
                    && calendar.before(getSecondHalfStartCalendar());
        }
    }

    public static boolean isSecondHalf() {
        Calendar calendar = Calendar.getInstance();
        if (!isWeekday()) {
            return false;
        } else {
            return calendar.after(getSecondHalfStartCalendar())
                    && calendar.before(getSecondHalfEndCalendar());
        }
    }

    public static boolean afterClosed() {
        Calendar calendar = Calendar.getInstance();
        if (!isWeekday()) {
            return false;
        } else {
            return calendar.after(getSecondHalfEndCalendar());
        }
    }

    private static Calendar getCalendar(String timeString) {
        Calendar calendar = Calendar.getInstance();
        String dateTimeString = (Utility.getCalendarDateString(calendar) + " "
                + timeString);
        return Utility.getCalendar(
                dateTimeString,
                Utility.CALENDAR_DATE_TIME_FORMAT
        );
    }

    public static Calendar getFirstHalfStartCalendar() {
        return getCalendar(FIRST_HALF_START_TIME);
    }

    public static Calendar getFirstHalfEndCalendar() {
        return getCalendar(FIRST_HALF_END_TIME);
    }

    public static Calendar getSecondHalfStartCalendar() {
        return getCalendar(SECOND_HALF_START_TIME);
    }

    public static Calendar getSecondHalfEndCalendar() {
        return getCalendar(SECOND_HALF_END_TIME);
    }
}

File: app/src/main/java/com/android/orion/utility/Market.kt
package com.android.orion.utility

import android.text.TextUtils
import java.util.*

object Market {
    private const val FIRST_HALF_START_TIME = "09:30:00"
    private const val FIRST_HALF_END_TIME = "11:30:00"
    private const val SECOND_HALF_START_TIME = "13:00:00"
    const val SECOND_HALF_END_TIME = "15:00:00"

    private const val START_IN_MINUTES = 9 * 60 + 30 //FIRST_HALF_START_TIME
    private const val LUNCH_TIME_IN_MINUTES =
        1 * 60 + 30 //from FIRST_HALF_END_TIME to SECOND_HALF_START_TIME

    private fun getMinutesOfToday(): Int {
        val calendar = Calendar.getInstance()
        return (calendar[Calendar.HOUR_OF_DAY] * 60
                + calendar[Calendar.MINUTE])
    }

    @JvmStatic
    fun getScheduleMinutes(): Int {
        return if (!isWeekday()) {
            0
        } else if (isFirstHalf()) {
            getMinutesOfToday() - START_IN_MINUTES
        } else if (isSecondHalf()) {
            getMinutesOfToday() - (START_IN_MINUTES
                    + LUNCH_TIME_IN_MINUTES)
        } else {
            0
        }
    }

    @JvmStatic
    fun isWeekday(): Boolean {
        val calendar: Calendar = Calendar.getInstance()
        val dayOfWeek = calendar[Calendar.DAY_OF_WEEK]
        return dayOfWeek > Calendar.SUNDAY && dayOfWeek < Calendar.SATURDAY
    }

    @JvmStatic
    fun isOutofDate(modified: String): Boolean {
        var result = false
        if (TextUtils.isEmpty(modified)) {
            result = true
        }
        if (!modified.contains(Utility.getCalendarDateString(Calendar.getInstance()))) {
            result = true
        }
        return result
    }

    @JvmStatic
    fun isTradingHours(): Boolean {
        return isFirstHalf() || isSecondHalf()
    }

    @JvmStatic
    fun beforeOpen(): Boolean {
        val calendar: Calendar = Calendar.getInstance()
        return if (!isWeekday()) {
            false
        } else calendar.before(getFirstHalfStartCalendar())
    }

    @JvmStatic
    fun isFirstHalf(): Boolean {
        val calendar: Calendar = Calendar.getInstance()
        return if (!isWeekday()) {
            false
        } else calendar.after(getFirstHalfStartCalendar())
                && calendar.before(getFirstHalfEndCalendar())
    }

    @JvmStatic
    fun isLunchTime(): Boolean {
        val calendar: Calendar = Calendar.getInstance()
        return if (!isWeekday()) {
            false
        } else calendar.after(getFirstHalfEndCalendar())
                && calendar.before(getSecondHalfStartCalendar())
    }

    @JvmStatic
    fun isSecondHalf(): Boolean {
        val calendar: Calendar = Calendar.getInstance()
        return if (!isWeekday()) {
            false
        } else calendar.after(getSecondHalfStartCalendar())
                && calendar.before(getSecondHalfEndCalendar())
    }

    @JvmStatic
    fun afterClosed(): Boolean {
        val calendar: Calendar = Calendar.getInstance()
        return if (!isWeekday()) {
            return false
        } else calendar.after(getSecondHalfEndCalendar())
    }

    fun getCalendar(timeString: String): Calendar {
        val calendar: Calendar = Calendar.getInstance()
        val dateTimeString = (Utility.getCalendarDateString(calendar) + " "
                + timeString)
        return Utility.getCalendar(
            dateTimeString,
            Utility.CALENDAR_DATE_TIME_FORMAT
        )
    }

    @JvmStatic
    fun getFirstHalfStartCalendar(): Calendar {
        return getCalendar(FIRST_HALF_START_TIME)
    }

    @JvmStatic
    fun getFirstHalfEndCalendar(): Calendar {
        return getCalendar(
            FIRST_HALF_END_TIME
        )
    }

    @JvmStatic
    fun getSecondHalfStartCalendar(): Calendar {
        return getCalendar(
            SECOND_HALF_START_TIME
        )
    }

    @JvmStatic
    fun getSecondHalfEndCalendar(): Calendar {
        return getCalendar(
            SECOND_HALF_END_TIME
        )
    }
}